crates/rpc/src/v02/method/get_nonce.rs

```diff
@@ -45,27 +48,151 @@ pub async fn get_nonce(
+            StarknetBlocksBlockId::Number(block_number) => {
+                // check that block exists
+                let latest = StarknetBlocksTable::get_latest_number(&tx)
+                    .context("Querying latest block number")?
+                    .ok_or(GetNonceError::BlockNotFound)?;
+                if block_number > latest {
+                    return Err(GetNonceError::BlockNotFound);
+                }
+
+                match get_nonce_at_block(&tx, contract_address, block_number)? {
+                    Some(nonce) => Ok(nonce),
+                    None => {
+                        database::contract_exists_at_block(&tx, contract_address, block_number)?
+                            .then_some(ContractNonce::default())
+                            .ok_or(GetNonceError::ContractNotFound)
+                    }
+                }
+            }
+            StarknetBlocksBlockId::Hash(block_hash) => {
+                // Get the block number from the hash.
+                let block_number = StarknetBlocksTable::get_number(&tx, block_hash)
+                    .context("Fetching block number")?
+                    .ok_or(GetNonceError::BlockNotFound)?;
+
+                match get_nonce_at_block(&tx, contract_address, block_number)? {
+                    Some(nonce) => Ok(nonce),
+                    None => {
+                        database::contract_exists_at_block(&tx, contract_address, block_number)?
+                            .then_some(ContractNonce::default())
+                            .ok_or(GetNonceError::ContractNotFound)
+                    }
+                }
+            }
+            StarknetBlocksBlockId::Latest => {
+                match database::get_nonce_at_latest(&tx, contract_address)? {
+                    Some(nonce) => Ok(nonce),
+                    None => database::contract_exists_at_latest(&tx, contract_address)?
+                        .then_some(ContractNonce::default())
+                        .ok_or(GetNonceError::ContractNotFound),
+                }
+            }
+        }?;
 
         Ok(GetNonceOutput(nonce))
     });
     jh.await.context("Database read panic or shutting down")?
 }
 
+mod database {
+    use pathfinder_common::StarknetBlockNumber;
+    use rusqlite::{named_params, OptionalExtension, Transaction};
+
+    use super::*;
+
+    pub fn get_nonce_at_latest(
+        tx: &Transaction,
+        contract_address: ContractAddress,
+    ) -> anyhow::Result<Option<ContractNonce>> {
+        tx.query_row(
+            r"SELECT felt FROM felts JOIN
+                (
+                    SELECT nonce as nonce_id FROM nonce_updates
+                        JOIN (SELECT id as contract_id FROM felts WHERE felt = :contract_address)
+                        WHERE nonce_updates.contract_address = contract_id
+                        ORDER BY nonce_updates.block_number DESC
+                        LIMIT 1
+                )
+                WHERE felts.id = nonce_id",
```

Comment:
These get complicated pretty quickly, but its not so bad so long as you remember that you are in index LUT mode.

Reply:
No longer applicable as we are not going to use interning.